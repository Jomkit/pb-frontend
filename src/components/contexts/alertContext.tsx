import { createContext } from "react";

const alertContext = createContext({} as {setAlertMessage: Function, setAlertOn: Function});

export default alertContext