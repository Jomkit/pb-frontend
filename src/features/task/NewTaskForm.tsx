import * as Yup from 'yup';
import userContext from '../../components/contexts/userContext';
import { useContext } from 'react';
import Api from '../../api';
import { useNavigate } from 'react-router-dom';
import { ErrorMessage, Field, Form, Formik } from 'formik';
import { IProject } from '../../types';
import alertContext from '../../components/contexts/alertContext';


const NewTaskForm = ({projects, closePopup}: {projects: IProject[], closePopup: () => void}) => {
    const navigate = useNavigate();
    const initialTaskValues = {
        projectId: "",
        name: '',
    }
    
    const currUser = useContext(userContext);
    const { setAlertMessage, setAlertOn } = useContext(alertContext);
    
    const newTaskSchema = Yup.object().shape({
        projectId: Yup.string().required().oneOf(projects.map( project => project.id!)),
        name: Yup.string().required(),
    })

    const handleSubmit = async (values: {projectId: string, name: string}) => {
        try{
            const {projectId, name} = values;
            console.log("values", values);
            await Api.createTask(projectId as string, {userId: currUser.id!, data: {name: name}});
            setAlertMessage(`Task ${name} created!`);
            setAlertOn(true);
            navigate('/');

        }catch(err: any){
            // Temporary fix for Clockify API restriction on unique names
            console.error("Error detected in NewTaskForm:", (err[0] || err));
            setAlertMessage("Task name already exists in app, please try another name");
            setAlertOn(true);

            closePopup();
        }
    }
    
  return (
    <div className='m-8 space-y-3'>
        <h2>Create New Task</h2>
        
        <Formik
        initialValues={initialTaskValues}
        validationSchema={newTaskSchema}
        onSubmit={handleSubmit}>
            <Form className='flex flex-col space-y-5'>
                <div>
                    <Field className="form-field1 focus:outline-none focus:shadow-outline" as="select" aria-label="projectId" name="projectId">
                        <option value="">Select Project</option>
                        {projects.map( (project) => (
                            <option key={project.id} value={project.id}>{project.name}</option>
                        ))}
                    </Field>
                    <ErrorMessage className='text-red-500' name="projectId" component="div" />
                </div>

                <div>
                    <Field className="form-field1 focus:outline-none focus:shadow-outline" name="name" type="text" placeholder="Task Name" />
                    <ErrorMessage className='text-red-500' name="name" component="div" />
                </div>
                
                <button className='btn bg-green-400' type="submit">Submit</button>
            </Form>
        </Formik>
    </div>
  )
}

export default NewTaskForm