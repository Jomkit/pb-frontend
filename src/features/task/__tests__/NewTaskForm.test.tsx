import { render } from "@testing-library/react"
import { MemoryRouter, useNavigate } from "react-router-dom"
import NewTaskForm from "../NewTaskForm"
import userEvent from "@testing-library/user-event"

import alertContext from "../../../components/contexts/alertContext";

vi.mock("../../../api");

const mockedClosePopup = vi.fn();
const mockedSetAlertMessage = vi.fn();
const mockedSetAlertOn = vi.fn();

const testProjects = [
    {
        id: "1",
        name: "testProject 1",
        note: "testNote"
    },
    {
        id: "2",
        name: "testProject 2",
        note: "testNote"
    },
    {
        id: "3",
        name: "testProject 3",
        note: "testNote"
    }
]

it("Should render without crashing", () => {
    render(
        <MemoryRouter>
            <NewTaskForm closePopup={mockedClosePopup} projects={testProjects} />
        </MemoryRouter>
    )
})

it("Should match snapshot", () => {
    const { asFragment, getByText } = render(
        <MemoryRouter>
            <NewTaskForm closePopup={mockedClosePopup} projects={testProjects} />
        </MemoryRouter>
    )

    expect(getByText("Create New Task")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
})

    it("should redirect to same page when successfully submitting form", async () => {
        const user = userEvent.setup();
        vi.mock("react-router-dom", async (importOriginal) => {
            const mod = await importOriginal<typeof import("react-router-dom")>();
            return {
                ...mod,
                useNavigate: vi.fn(() => vi.fn())
            };
        })

        const { getByPlaceholderText, getByText, getByRole } = render(
            <MemoryRouter>
                <alertContext.Provider value={{setAlertMessage: mockedSetAlertMessage, setAlertOn: mockedSetAlertOn}} >
                    <NewTaskForm closePopup={mockedClosePopup} projects={testProjects} />
                </alertContext.Provider>
            </MemoryRouter>
        );

        const submitBtn = getByText("Submit");

        const projectSelect = getByRole("combobox", {name: "projectId"});
        const nameInput = getByPlaceholderText("Task Name");

        await user.selectOptions(projectSelect, ["1"]);
        await user.type(nameInput, "testName");
        await user.click(submitBtn);
        expect(useNavigate).toHaveBeenCalled();
        expect(mockedSetAlertOn).toHaveBeenCalled();
        expect(mockedSetAlertMessage).toHaveBeenCalled();
    })