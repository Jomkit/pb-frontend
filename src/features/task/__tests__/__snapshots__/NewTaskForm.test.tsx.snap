// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="m-8 space-y-3"
  >
    <h2>
      Create New Task
    </h2>
    <form
      action="#"
      class="flex flex-col space-y-5"
    >
      <div>
        <select
          aria-label="projectId"
          class="form-field1 focus:outline-none focus:shadow-outline"
          name="projectId"
        >
          <option
            value=""
          >
            Select Project
          </option>
          <option
            value="1"
          >
            testProject 1
          </option>
          <option
            value="2"
          >
            testProject 2
          </option>
          <option
            value="3"
          >
            testProject 3
          </option>
        </select>
      </div>
      <div>
        <input
          class="form-field1 focus:outline-none focus:shadow-outline"
          name="name"
          placeholder="Task Name"
          type="text"
          value=""
        />
      </div>
      <button
        class="btn bg-green-400"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
